// @pulumi/azuread/applicationOptionalClaims.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApplicationOptionalClaims = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("./utilities");
/**
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as azuread from "@pulumi/azuread";
 *
 * const example = new azuread.ApplicationRegistration("example", {displayName: "example"});
 * const exampleApplicationOptionalClaims = new azuread.ApplicationOptionalClaims("example", {
 *     applicationId: example.id,
 *     accessTokens: [
 *         {
 *             name: "myclaim",
 *         },
 *         {
 *             name: "otherclaim",
 *         },
 *     ],
 *     idTokens: [{
 *         name: "userclaim",
 *         source: "user",
 *         essential: true,
 *         additionalProperties: ["emit_as_roles"],
 *     }],
 *     saml2Tokens: [{
 *         name: "samlexample",
 *     }],
 * });
 * ```
 *
 * ## Import
 *
 * Application Optional Claims can be imported using the object ID of the application, in the following format.
 *
 * ```sh
 * $ pulumi import azuread:index/applicationOptionalClaims:ApplicationOptionalClaims example /applications/00000000-0000-0000-0000-000000000000
 * ```
 */
class ApplicationOptionalClaims extends pulumi.CustomResource {
    /**
     * Get an existing ApplicationOptionalClaims resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ApplicationOptionalClaims(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ApplicationOptionalClaims.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ApplicationOptionalClaims.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accessTokens"] = state ? state.accessTokens : undefined;
            resourceInputs["applicationId"] = state ? state.applicationId : undefined;
            resourceInputs["idTokens"] = state ? state.idTokens : undefined;
            resourceInputs["saml2Tokens"] = state ? state.saml2Tokens : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.applicationId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'applicationId'");
            }
            resourceInputs["accessTokens"] = args ? args.accessTokens : undefined;
            resourceInputs["applicationId"] = args ? args.applicationId : undefined;
            resourceInputs["idTokens"] = args ? args.idTokens : undefined;
            resourceInputs["saml2Tokens"] = args ? args.saml2Tokens : undefined;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(ApplicationOptionalClaims.__pulumiType, name, resourceInputs, opts);
    }
}
exports.ApplicationOptionalClaims = ApplicationOptionalClaims;
/** @internal */
ApplicationOptionalClaims.__pulumiType = 'azuread:index/applicationOptionalClaims:ApplicationOptionalClaims';
